### DISCUSSION TITLE
好奇具体思路是什么

### DISCUSSION THREAD

**Discussion by xycabcd** [baulk/baulk#66] [General]
这几天想写一个macOS的包管理器，想请教一下一个包管理器的具体思路是什么 （不是大体）

**Comment by xycabcd** [comment]
感激不尽

我现在完全没有头绪啊

**Reply by xycabcd** [reply]
如果能有伪代码啥的就更好了

**Comment by fcharlie** [comment]
其实原理你可以学习一下 brew 的思路，在 macOS brew 都已经成为事实的标准了。
以 baulk 为例，baulk 需要建立一个 bucket，也就是软件源，记录了软件的版本，下载地址，文件哈希，格式，一些其他信息，包管理器需要做的就是更新软件源信息，用户运行安装指令时，下载文件，解压，安装（绿色版解压好即可），然后实现各种各样的命令。

**Reply by xycabcd** [reply]
感谢。
brew看过，ruby看不懂。
不过头绪已经开始出现了。
应该最终可以"build from scratch"吧

**Reply by xycabcd** [reply]
> 下载文件，解压，安装（绿色版解压好即可)  

Unix 不一样，大部分命令行软件都需要自己编译（因为有依赖，要把依赖的路径编译到程序里），这里面还有很多坑（见：https://github.com/orgs/Homebrew/discussions/5405#discussioncomment-9566669  ）。这里brew采用的是自己预先编译好调试好给用户下载，但是就只能安装在那一个位置。MacPorts现在使用不知道什么方法解决了这些问题。当然，我天天用Homebrew源码编译也没遇到这种问题。
但是我感觉这些都是后面精修要做的，前面就是先写个框架出来再说。

**Reply by xycabcd** [reply]
想再问一句，就是Windows下面有没有需要本地编译的软件，如果有的话，又是怎样进行的呢？
(因为Unix下面几乎所有软件都要编译)

**Reply by fcharlie** [reply]
> 想再问一句，就是Windows下面有没有需要本地编译的软件，如果有的话，又是怎样进行的呢？ (因为Unix下面几乎所有软件都要编译)

维护一个构建仓库，走 github release

**Reply by xycabcd** [reply]
所以说这些都是绿色软件吗。
只可惜Unix下面不行啊。Unix下面喜欢弄依赖别的软件，然后别的软件的路径是要编译的时候直接写到bin里头的。只能本地编译了。

**Reply by fcharlie** [reply]
> 所以说这些都是绿色软件吗。 只可惜Unix下面不行啊。Unix下面喜欢弄依赖别的软件，然后别的软件的路径是要编译的时候直接写到bin里头的。只能本地编译了。

Unix 也可以做到，只是稍微麻烦，静态编译 or 指定 --prefix，还可以修改 rpath/runpath，参考我好些年前从 cmake 抽出来的实现 cmchrpath（就是修改 ELF rpath 设置）：https://github.com/fcharlie/cmchrpath

**Reply by xycabcd** [reply]
当然，直接本地编译是最稳妥的方法，大部分都用默认选项就行（除了路径啥的），特别是对我这种经验不足的比较不容易出毛病。
